### PR TITLE
LPS-150912 Avoid sharing database connections between threads during upgrades when DB Partitioning is enabled

### PR DESCRIPTION
**Pull request by LuisOrtiz89**
LPS-150912

### CODE REVIEW

**Comment by LuisOrtiz89** [comment]
Hi @shuyangzhou, new version with the JDK Dynamic Proxy as indicated. This one looks much better.
Could we consider using the proxy always on upgrade and verify processes, being the database partitioning enabled or not? I am thinking in this ticket https://issues.liferay.com/browse/LPS-151414, where we are having issues on SQL Server due to the processConcurrently method, which is independent on the Database Partitioning. Using the proxy would prevent this cases. If not possible, we should review the processConcurrently method, as probably the issue in the ticket will happen again in other upgrade processes.

**Comment by shuyangzhou** [comment]
@LuisOrtiz89 please add a test for this, I need a way to debug through your code.

**Comment by LuisOrtiz89** [comment]
Hi @shuyangzhou, I added some tests, please let me know if this is OK or you miss something more.

**Comment by shuyangzhou** [comment]
ci:test:relevant

**Comment by LuisOrtiz89** [comment]
Hi @shuyangzhou, from the tests I saw there was some problem with the static initialization of a couple new variables, so I rewrite that part and make new commits. I am going to try pass the tests again.

**Comment by LuisOrtiz89** [comment]
ci:test:sf

**Comment by LuisOrtiz89** [comment]
ci:test:relevant

File: portal-kernel/src/com/liferay/portal/kernel/upgrade/UpgradeProcess.java [hunk 1]
```diff
@@ -86,7 +90,30 @@ public void upgrade() throws UpgradeException {
...
+		try {
+			if (partitionConcurrencyEnabled) {
+				connection = (Connection)ProxyUtil.newProxyInstance(
+					ClassLoader.getSystemClassLoader(),
+					new Class<?>[] {Connection.class},
+					new ConnectionThreadProxyHandler());
+			}
+			else {
+				connection = getConnection();
+			}
```
Comment (shuyangzhou) [inline portal-kernel/src/com/liferay/portal/kernel/upgrade/UpgradeProcess.java:115]: @LuisOrtiz89 this section looks problematic. If you see the body of "getConnection()", it is doing special handling for external datasource, but once you go into the parallel branch, it always uses portal's default datasource.
All the logic for connection getting (including possible paralllel proxy wrapping) should be moved into "getConnection()" method, and the method should be promoted to BaseDBProcess in order to share with VerifyProcess.
Make sure the external datasources are properly handled even with parallel activation.

File: portal-kernel/src/com/liferay/portal/kernel/upgrade/UpgradeProcess.java [hunk 2]
```diff
@@ -115,6 +142,17 @@ public void upgrade() throws UpgradeException {
 			throw new UpgradeException(throwable);
 		}
 		finally {
+			if (connection != null) {
```
Comment (shuyangzhou) [inline portal-kernel/src/com/liferay/portal/kernel/upgrade/UpgradeProcess.java:145]: @LuisOrtiz89 why is there a null check here? connection is never null in this method.

**Comment by shuyangzhou** [comment]
@LuisOrtiz89 please see the above comments, the external datasource issue must be fixed, otherwise we would have a broken upgrade when there are modules using foreign datasources.

**Pull request closed without merging**